### ISSUE TITLE
common arguments shared between sub-commands

### ISSUE THREAD

**Issue by kopp** [typed-argparse/typed-argparse#31]
Could you please extend the example for sub commands by providing an argument that can be passed to demo_app and is used for both foo and bar (for example `demo_app --verbose foo start` if that is possible (or a note that this is not possible)?

What I understand from the documentation is, that if I wanted to support a flag for all commands I would need to either make the different `*Args` classes subclasses of some class providing common options or repeat them for each subcommands and handle them separately for each sub command (i.e. call `demo_app foo start --verbose`).

**Comment by bluenote10** [comment]
Yes, I wanted to add a specific documentation example for that.

In short: Yes, just introduce a `CommonArgs(TypedArg)` from which your individual command args inherit. And then you have both options:
- If you **do not** pass the `CommonArgs` as the `common_args=...` argument to the subparser group, the common arg will be added to each command indivdually, i.e., usage becomes `<myapp> foo --verbose` and `<myapp> bar --verbose`.
- If you **do** pass the `CommonArgs` as the `common_args=...` argument to the subparser group, the common arg will be added at the parent level, i.e., usage becomes `<myapp> --verbose foo` and `<myapp> --verbose bar`.

Note: It's not strictly necessary to use inheritance, but this is obviously the simplest way to avoid repeating fields...

**Comment by pdc1** [comment]
I tried this (second option, **do** pass `CommonArgs`) using your samples, and it works to a point. If you run the command with no arguments, you get the full usage:

```bash
$ python3 tap-example.py
usage: tap-example.py [-h] --my-arg MY_ARG [--number-a NUMBER_A]
                      [--number-b NUMBER_B] [--verbose] --names
                      [NAMES [NAMES ...]]
                      {foo,bar} ...
tap-example.py: error: the following arguments are required: --my-arg, --names, <sub-command>
```

However, if a subcommand is given, the usage does not include the common arguments:

```bash
$ python3 tap-example.py foo
usage: tap-example.py foo [-h] {start,stop} ...
tap-example.py foo: error: the following arguments are required: <sub-sub-command>
```

Also note that the usage in the second case does not include the required common arguments. If you satisfy all of the arguments for the `foo` subcommand, THEN the parser will tell you about the required common arguments.

I know there's a lot going on here behind the scenes, so I don't mean to be critical but I wanted to point this out as a caveat. I use the low-level API to give fine control over things like this, so thanks for providing both approaches.

**Comment by bluenote10** [comment]
> However, if a subcommand is given, the usage does not include the common arguments:

Isn't that just how subcommand parsing works typically? At least, it is just how argparse works, and many other popular behave this way, right? For instance, `git pull --help` or `pip install --help` also seem to not report "top level" arguments, neither does the misusage of them.

**Comment by pdc1** [comment]
> Isn't that just how subcommand parsing works typically? 

Oh right, good point, that seems reasonable. Never mind that part :)

**Comment by bluenote10** [comment]
Should be closed by #37
